xclim/core/options.py

```diff
@@ -0,0 +1,169 @@
+"""Global or contextual options for xclim, similar to xarray.set_options"""
+import logging
+from inspect import signature
+from pathlib import Path
+from warnings import warn
+
+from boltons.funcutils import wraps
+
+from xclim.core.utils import ValidationError
+
+logging.captureWarnings(True)
+
+
+METADATA_LOCALES = "metadata_locales"
+VALIDATE_INPUTS = "validate_inputs"
+CHECK_MISSING = "check_missing"
+MISSING_OPTIONS = "missing_options"
+
+MISSING_METHODS = {}
+
+OPTIONS = {
+    METADATA_LOCALES: [],
+    VALIDATE_INPUTS: "warn",
+    CHECK_MISSING: "any",
+    MISSING_OPTIONS: MISSING_METHODS,
+}
+
+_VALIDATION_OPTIONS = frozenset(["log", "warn", "raise"])
+
+
+def _valid_locales(locales):
+    from xclim.locales import get_best_locale
+
+    return all(
+        [
+            (isinstance(locale, str) and get_best_locale(locale) is not None)
+            or (
+                not isinstance(locale, str)
+                and isinstance(locale[0], str)
+                and (Path(locale[1]).is_file() or isinstance(locale[1], dict))
+            )
+            for locale in locales
+        ]
+    )
+
+
+def _valid_missing_options(mopts):
+    for meth, opts in mopts.items():
+        if (
+            meth not in MISSING_METHODS  # Method must be registered
+            # All options must exist
+            or any([opt not in MISSING_METHODS[meth] for opt in opts.keys()])
+            # Method option validator, if it exists, must pass
+            or (
+                hasattr(MISSING_METHODS[meth]["_cls"], "validate")
+                and not MISSING_METHODS[meth]["_cls"].validate(
+                    **{k: v for k, v in opts.items() if k != "_cls"}
+                )
+            )
+            # Must not overwrite cls
+            or "_cls" in opts
+        ):
+            return False
+    return True
+
+
+_VALIDATORS = {
+    METADATA_LOCALES: _valid_locales,
+    VALIDATE_INPUTS: _VALIDATION_OPTIONS.__contains__,
+    CHECK_MISSING: MISSING_METHODS.__contains__,
+    MISSING_OPTIONS: _valid_missing_options,
+}
+
+
+def _set_missing_options(mopts):
+    for meth, opts in mopts.items():
+        MISSING_METHODS[meth].update(opts)
+
+
+_SETTERS = {MISSING_OPTIONS: _set_missing_options}
+
+
+def register_missing_method(name, validator=None):
+    def _register_missing_method(cls):
+        sig = signature(cls.is_missing)
+        opts = {
+            name: param.default
+            for name, param in sig.parameters.items()
+            if name not in ["self", "null", "count"]
+        }
+        MISSING_METHODS[name] = {"_cls": cls, **opts}
```

Comment:
I suggest `register` fills two dict: 
- {name: class}
- {name: dict(**missing arguments)}

Having a `_cls` item forces you to do all kinds of special casing afterward.